File: openrndr-jvm/openrndr-adopted-artifex/src/main/java/io/lacuna/artifex/Box3.java
package io.lacuna.artifex;

import static io.lacuna.artifex.Vec.vec;
import static java.lang.Double.NaN;

/**
 * @author ztellman
 */
public class Box3 extends Box<Vec3, Box3> {

  public static final Box3 EMPTY = new Box3(vec(NaN, NaN, NaN), vec(NaN, NaN, NaN));
  
  public final double lx, ly, lz, ux, uy, uz;
  
  Box3(double ax, double ay, double az, double bx, double by, double bz) {
    if (ax < bx) {
      this.lx = ax;
      this.ux = bx;
    } else {
      this.ux = ax;
      this.lx = bx;
    }

    if (ay < by) {
      this.ly = ay;
      this.uy = by;
    } else {
      this.uy = ay;
      this.ly = by;
    }

    if (az < bz) {
      this.lz = az;
      this.uz = bz;
    } else {
      this.uz = az;
      this.lz = bz;
    }
  }

  public Box3(Vec3 a, Vec3 b) {
    this(a.x, a.y, a.z, b.x, b.y, b.z);
  }

  public Box2 box2() {
    return new Box2(lx, ly, ux, uz);
  }

  @Override
  protected Box3 construct(Vec3 a, Vec3 b) {
    return new Box3(a, b);
  }

  @Override
  protected Box3 empty() {
    return EMPTY;
  }

  @Override
  public Vec3 lower() {
    return new Vec3(lx, ly, lz);
  }

  @Override
  public Vec3 upper() {
    return new Vec3(ux, uy, uz);
  }

  @Override
  public boolean isEmpty() {
    return this == EMPTY;
  }
}

File: openrndr-kartifex/src/commonMain/kotlin/Box3.kt
package org.openrndr.kartifex


/**
 * Represents a three-dimensional axis-aligned box defined by two points in 3D space.
 *
 * @constructor Initializes the box with the specified lower (`a`) and upper (`b`) points.
 * Internally determines the correct lower and upper bounds along each axis.
 * This ensures that, regardless of input order, the box is represented correctly.
 *
 * @property lx The lower bound of the box along the x-axis.
 * @property ly The lower bound of the box along the y-axis.
 * @property lz The lower bound of the box along the z-axis.
 * @property ux The upper bound of the box along the x-axis.
 * @property uy The upper bound of the box along the y-axis.
 * @property uz The upper bound of the box along the z-axis.
 *
 * @constructor A secondary constructor that initializes the box using two `Vec3` objects
 * representing the lower and upper bounds.
 */
class Box3 internal constructor(ax: Double, ay: Double, az: Double, bx: Double, by: Double, bz: Double) :
    Box<Vec3, Box3>() {
    val lx: Double
    val ly: Double
    val lz: Double
    val ux: Double
    val uy: Double
    val uz: Double

    constructor(a: Vec3, b: Vec3) : this(a.x, a.y, a.z, b.x, b.y, b.z)

    fun box2(): Box2 {
        return Box2(lx, ly, ux, uz)
    }

    override fun construct(a: Vec3, b: Vec3): Box3 {
        return Box3(a, b)
    }

    override fun empty(): Box3 {
        return EMPTY
    }

    override fun lower(): Vec3 {
        return Vec3(lx, ly, lz)
    }

    override fun upper(): Vec3 {
        return Vec3(ux, uy, uz)
    }

    override val isEmpty: Boolean
        get() = this === EMPTY

    companion object {
        val EMPTY = Box3(
            Vec3(Double.NaN, Double.NaN, Double.NaN),
            Vec3(Double.NaN, Double.NaN, Double.NaN)
        )
    }

    init {
        if (ax < bx) {
            lx = ax
            ux = bx
        } else {
            ux = ax
            lx = bx
        }
        if (ay < by) {
            ly = ay
            uy = by
        } else {
            uy = ay
            ly = by
        }
        if (az < bz) {
            lz = az
            uz = bz
        } else {
            uz = az
            lz = bz
        }
    }
}
